<!doctype html>
<!--
  Minimal Mistakes Jekyll Theme 4.13.0 by Michael Rose
  Copyright 2013-2018 Michael Rose - mademistakes.com | @mmistakes
  Free for personal and commercial use under the MIT license
  https://github.com/mmistakes/minimal-mistakes/blob/master/LICENSE
-->
<html lang="en" class="no-js">
  <head>
    <meta charset="utf-8">

<!-- begin _includes/seo.html --><title>Learning How to Run with Genetic Algorithms - Brandon Da Silva</title>
<meta name="description" content="Reinforcement Learning, Neural Networks, Genetic Algorithm">



<meta property="og:type" content="article">
<meta property="og:locale" content="en_US">
<meta property="og:site_name" content="Brandon Da Silva">
<meta property="og:title" content="Learning How to Run with Genetic Algorithms">
<meta property="og:url" content="http://localhost:4000/genetic-algorithm/">


  <meta property="og:description" content="Reinforcement Learning, Neural Networks, Genetic Algorithm">



  <meta property="og:image" content="http://localhost:4000/images/robotEvolution.jpg">





  <meta property="article:published_time" content="2018-11-18T00:00:00-05:00">





  

  


<link rel="canonical" href="http://localhost:4000/genetic-algorithm/">







  <script type="application/ld+json">
    {
      "@context": "http://schema.org",
      "@type": "Person",
      "name": "Brandon Da Silva",
      "url": "http://localhost:4000",
      "sameAs": null
    }
  </script>







<!-- end _includes/seo.html -->


<link href="/feed.xml" type="application/atom+xml" rel="alternate" title="Brandon Da Silva Feed">

<!-- http://t.co/dKP3o1e -->
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<script>
  document.documentElement.className = document.documentElement.className.replace(/\bno-js\b/g, '') + ' js ';
</script>

<!-- For all browsers -->
<link rel="stylesheet" href="/assets/css/main.css">

<!--[if lte IE 9]>
  <style>
    /* old IE unsupported flexbox fixes */
    .greedy-nav .site-title {
      padding-right: 3em;
    }
    .greedy-nav button {
      position: absolute;
      top: 0;
      right: 0;
      height: 100%;
    }
  </style>
<![endif]-->



    <!-- start custom head snippets -->

<!-- insert favicons. use https://realfavicongenerator.net/ -->

<!-- end custom head snippets -->

  </head>

  <body class="layout--single">

    <!--[if lt IE 9]>
<div class="notice--danger align-center" style="margin: 0;">You are using an <strong>outdated</strong> browser. Please <a href="https://browsehappy.com/">upgrade your browser</a> to improve your experience.</div>
<![endif]-->

    <div class="masthead">
  <div class="masthead__inner-wrap">
    <div class="masthead__menu">
      <nav id="site-nav" class="greedy-nav">
        <a class="site-title" href="/">Brandon Da Silva</a>
        <ul class="visible-links"><li class="masthead__menu-item">
              <a href="/about/" >About</a>
            </li><li class="masthead__menu-item">
              <a href="/machine-learning/" >Machine Learning</a>
            </li></ul>
        
        <button class="greedy-nav__toggle hidden" type="button">
          <span class="visually-hidden">Toggle menu</span>
          <div class="navicon"></div>
        </button>
        <ul class="hidden-links hidden"></ul>
      </nav>
    </div>
  </div>
</div>


    <div class="initial-content">
      
  











<div class="page__hero"
  style=" "
>
  
    <img src="/images/robotEvolution.jpg" alt="Learning How to Run with Genetic Algorithms" class="page__hero-image">
  
  
</div>





<div id="main" role="main">
  
  <div class="sidebar sticky">
  


<div itemscope itemtype="http://schema.org/Person">

  
    <div class="author__avatar">
      

      
        <img src="/images/bio-photo.jpg" alt="Brandon Da Silva" itemprop="image">
      
    </div>
  

  <div class="author__content">
    
      <h3 class="author__name" itemprop="name">Brandon Da Silva</h3>
    
    
      <p class="author__bio" itemprop="description">
        Machine Learning Blog
      </p>
    
  </div>

  <div class="author__urls-wrapper">
    <button class="btn btn--inverse">Follow</button>
    <ul class="author__urls social-icons">
      
        <li itemprop="homeLocation" itemscope itemtype="http://schema.org/Place">
          <i class="fas fa-fw fa-map-marker-alt" aria-hidden="true"></i> <span itemprop="name">Toronto, ON</span>
        </li>
      

      
        
          
        
          
        
          
        
          
        
          
        
          
        
      

      

      
        <li>
          <a href="mailto:brandasilva9@gmail.com">
            <meta itemprop="email" content="brandasilva9@gmail.com" />
            <i class="fas fa-fw fa-envelope-square" aria-hidden="true"></i> Email
          </a>
        </li>
      

      

      

      

      

      
        <li>
          <a href="https://www.linkedin.com/in/brandon-da-silva-a67496a7" itemprop="sameAs">
            <i class="fab fa-fw fa-linkedin" aria-hidden="true"></i> LinkedIn
          </a>
        </li>
      

      

      

      

      

      
        <li>
          <a href="https://github.com/brandinho" itemprop="sameAs">
            <i class="fab fa-fw fa-github" aria-hidden="true"></i> GitHub
          </a>
        </li>
      

      

      

      

      

      

      

      

      

      

      

      

      

      

      <!--
  <li>
    <a href="http://link-to-whatever-social-network.com/user/" itemprop="sameAs">
      <i class="fas fa-fw" aria-hidden="true"></i> Custom Social Profile Link
    </a>
  </li>
-->
    </ul>
  </div>
</div>

  
  </div>


  <article class="page" itemscope itemtype="http://schema.org/CreativeWork">
    <meta itemprop="headline" content="Learning How to Run with Genetic Algorithms">
    <meta itemprop="description" content="Reinforcement Learning, Neural Networks, Genetic Algorithm">
    <meta itemprop="datePublished" content="November 18, 2018">
    

    <div class="page__inner-wrap">
      
        <header>
          <h1 id="page-title" class="page__title" itemprop="headline">Learning How to Run with Genetic Algorithms
</h1>
          
            <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  8 minute read
</p>
          
        </header>
      

      <section class="page__content" itemprop="text">
        
        <h2 id="overview">Overview</h2>

<p>When most people think of Deep Reinforcement Learning, they probably think of Q-networks or policy gradients. Both of these methods require you to calculate derivatives and use gradient descent. In this post, we are going to explore a derivative-free method for optimizing a policy network. Specifically, we are going to be using a genetic algorithm on DeepMind’s <a href="https://arxiv.org/pdf/1801.00690.pdf">Control Suite</a> to allow the “cheetah” physical model to learn how to run. You can find the complete code on my <a href="https://github.com/brandinho/Genetic-Algorithm-Control-Suite">github repo</a>.</p>

<h2 id="genetic-algorithm-background">Genetic Algorithm Background</h2>

<p>Genetic algorithms (GAs) are inspired by natural selection, as put forth by Charles Darwin. The idea is that over generations, the heritable traits of a population change because of <em>mutation</em> and the concept of <em>survival of the fittest</em>.</p>

<p>Similar to natural selection, GAs iterate over multiple generations to evolve a population. The population in our case is going to consist of a bunch of neural network weights, which define our cheetah agents. You can think of each set of neural network weights as an individual agent in the population - usually called a chromosome or genotype. Chromosomes are usually encoded as binary strings, but since we want to optimize neural networks weights, we will adapt it for continuous numbers. Each neural network weight in our chromosome can be referred to as a gene. After iterating through all the generations, and continually improving the cheetah’s chromosome (its neural network weights), we hope that it learns how to run.</p>

<h3 id="initialization">Initialization</h3>

<p>To begin the process, we need to initialize our population of agents. We sample the initial neural network weights from a normal distribution with a scaling factor outlined in <a href="http://proceedings.mlr.press/v9/glorot10a/glorot10a.pdf">Glorot and Bengio’s paper</a>:</p>

<script type="math/tex; mode=display">Var[W^i] = \frac{2}{n_i + n_{i+1}}</script>

<p>where <script type="math/tex">W^i</script> refers to the weight matrix in the <script type="math/tex">i^\text{th}</script> layer, while <script type="math/tex">n_i</script> and <script type="math/tex">n_{i+1}</script> refer to the input and output dimensionality of that layer. Below you’ll see python code to implement the population initialization, where <code class="highlighter-rouge">scaling_factor</code> is a vector of variances calculated according to the equation above:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">population</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">multivariate_normal</span><span class="p">(</span><span class="n">mean</span> <span class="o">=</span> <span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">*</span><span class="n">scaling_factor</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">],</span>
                                             <span class="n">cov</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">diag</span><span class="p">(</span><span class="n">scaling_factor</span><span class="p">),</span>
                                             <span class="n">size</span> <span class="o">=</span> <span class="n">population_size</span><span class="p">)</span>  
</code></pre></div></div>

<h3 id="selection">Selection</h3>

<p>Now that we have a population, we can have the agents within the population compete against each other! The agents that are the most “fit” have the highest probability of passing their genes onto the next generation. We will define fitness as the cumulative reward of our agent over the span of an episode. As you might have guessed by the way we defined it, fitness refers to how good an agent is at performing the task we want it to learn. Those that are better at performing the task will have a better chance of being selected as parents to breed a new generation. There are two primary methods for parent selection - <strong>Roulette</strong> and <strong>Tournament</strong>.</p>

<p>The roulette method selects parents with a probability proportional to their fitness score. This is why it is also called <em>Fitness Proportionate Selection</em>.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="c1"># Roulette Wheel Selection
</span>  <span class="n">position</span> <span class="o">=</span> <span class="p">[]</span>
  <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">2</span><span class="p">):</span>
      <span class="n">random_number</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">uniform</span><span class="p">(</span><span class="n">low</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="n">high</span> <span class="o">=</span> <span class="n">scores_cumulsum</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
      <span class="n">position</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="nb">next</span><span class="p">(</span><span class="n">x</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">scores_cumulsum</span><span class="p">)</span> <span class="k">if</span> <span class="n">x</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">&gt;</span> <span class="n">random_number</span><span class="p">))</span>

  <span class="n">parent_1</span> <span class="o">=</span> <span class="n">population</span><span class="p">[</span><span class="n">population_index_sorted</span><span class="p">[</span><span class="n">position</span><span class="p">[</span><span class="mi">0</span><span class="p">]]]</span>
  <span class="n">parent_2</span> <span class="o">=</span> <span class="n">population</span><span class="p">[</span><span class="n">population_index_sorted</span><span class="p">[</span><span class="n">position</span><span class="p">[</span><span class="mi">1</span><span class="p">]]]</span>
</code></pre></div></div>

<p>The tournament method runs two tournaments in parallel with different subsets of the total population. The competitors for each tournament are chosen at random. The winners from each tournament are selected as the parents to breed the next generation.</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="c1"># Tournament Selection
</span>  <span class="n">k</span> <span class="o">=</span> <span class="n">population_size</span> <span class="o">//</span> <span class="mi">2</span>
  <span class="n">tournament_population</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">zeros</span><span class="p">((</span><span class="n">k</span><span class="p">,</span> <span class="mi">2</span><span class="p">))</span>
  <span class="n">total_competitors</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">population_size</span><span class="p">),</span> <span class="n">k</span> <span class="o">*</span> <span class="mi">2</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
  <span class="n">tournament_population</span><span class="p">[:,</span><span class="mi">0</span><span class="p">]</span> <span class="o">=</span> <span class="n">competition_scores</span><span class="p">[</span><span class="n">total_competitors</span><span class="p">[:</span><span class="n">k</span><span class="p">]]</span>
  <span class="n">tournament_population</span><span class="p">[:,</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">competition_scores</span><span class="p">[</span><span class="n">total_competitors</span><span class="p">[</span><span class="n">k</span><span class="p">:]]</span>

  <span class="n">parent_indexes</span> <span class="o">=</span> <span class="n">total_competitors</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">argmax</span><span class="p">(</span><span class="n">tournament_population</span><span class="p">,</span> <span class="n">axis</span> <span class="o">=</span> <span class="mi">0</span><span class="p">)</span> <span class="o">+</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">([</span><span class="mi">0</span><span class="p">,</span><span class="n">k</span><span class="p">])]</span>
  <span class="n">parent_1</span> <span class="o">=</span> <span class="n">population</span><span class="p">[</span><span class="n">parent_indexes</span><span class="p">[</span><span class="mi">0</span><span class="p">],]</span>
  <span class="n">parent_2</span> <span class="o">=</span> <span class="n">population</span><span class="p">[</span><span class="n">parent_indexes</span><span class="p">[</span><span class="mi">1</span><span class="p">],]</span>
</code></pre></div></div>

<h3 id="elitism">Elitism</h3>

<p>One thing we can do to improve performance in our algorithm is introduce the concept of elitism. This refers to the act of carrying over the most fit agents to the next generation without altering their chromosomes through crossover or mutation (which we will explore very soon). We do this because we always want to preserve the best agents from one generation to the next; it is not guaranteed that any of the children will be more fit than their parents.</p>

<h3 id="crossover">Crossover</h3>

<p>Now that we know how to select the parents from the population, let’s talk breeding! Crossover, also called recombination, takes the chromosomes of two parents and combines them to form children in the next generation. Here are a few ways you can combine two chromosomes:</p>

<p>The first and easiest way is to perform <strong>One Point</strong> crossover. You randomly select a partition in the chromosome, as indicated by the red line below. The child gets the left side of the partition from one parent and the right side from the other parent.</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/one_point_crossover.png" alt="alt" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">partition</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">parent_1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
  <span class="c1"># Select which parent will be the "left side"
</span>  <span class="k">if</span> <span class="n">which_parent</span> <span class="o">==</span> <span class="s">"Parent 1"</span><span class="p">:</span>
      <span class="n">child</span> <span class="o">=</span> <span class="n">parent_1</span>
      <span class="n">child</span><span class="p">[</span><span class="n">partition</span><span class="p">:]</span> <span class="o">=</span> <span class="n">parent_2</span><span class="p">[</span><span class="n">partition</span><span class="p">:]</span>
  <span class="k">elif</span> <span class="n">which_parent</span> <span class="o">==</span> <span class="s">"Parent 2"</span><span class="p">:</span>
      <span class="n">child</span> <span class="o">=</span> <span class="n">parent_2</span>
      <span class="n">child</span><span class="p">[</span><span class="n">partition</span><span class="p">:]</span> <span class="o">=</span> <span class="n">parent_1</span><span class="p">[</span><span class="n">partition</span><span class="p">:]</span>
</code></pre></div></div>

<p>Building on the previous method is <strong>Two Point</strong> crossover. This is conceptually the same, except you randomly select two points, which serve as a lower and upper bound. The child gets the elements outside of the bounds from one parent, and the elements within the bounds from the other parent.</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/two_point_crossover.png" alt="alt" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">lower_limit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">parent_1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
  <span class="n">upper_limit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">lower_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">parent_1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
  <span class="c1"># Select which parent will be the "outside bounds"
</span>  <span class="k">if</span> <span class="n">which_parent</span> <span class="o">==</span> <span class="s">"Parent 1"</span><span class="p">:</span>
      <span class="n">child</span> <span class="o">=</span> <span class="n">parent_1</span>
      <span class="n">child</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent_2</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
  <span class="k">elif</span> <span class="n">which_parent</span> <span class="o">==</span> <span class="s">"Parent 2"</span><span class="p">:</span>
      <span class="n">child</span> <span class="o">=</span> <span class="n">parent_2</span>
      <span class="n">child</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">parent_1</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span>
</code></pre></div></div>

<p>Unlike the previous two methods, which required the swapped genes to be in a sequence, the <strong>Uniform</strong> crossover does not. Rather, it randomly selects, with a uniform distribution, the indexes to be swapped during crossover.</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/uniform_crossover.png" alt="alt" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">random_sequence</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">parent_1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">parent_1</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="n">replace</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
  <span class="k">if</span> <span class="n">which_parent</span> <span class="o">==</span> <span class="s">"Parent 1"</span><span class="p">:</span>
      <span class="n">child</span> <span class="o">=</span> <span class="n">parent_1</span>
      <span class="n">child</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">random_sequence</span><span class="p">)]</span> <span class="o">=</span> <span class="n">parent_2</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">random_sequence</span><span class="p">)]</span>
  <span class="k">elif</span> <span class="n">which_parent</span> <span class="o">==</span> <span class="s">"Parent 2"</span><span class="p">:</span>
      <span class="n">child</span> <span class="o">=</span> <span class="n">parent_2</span>
      <span class="n">child</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">random_sequence</span><span class="p">)]</span> <span class="o">=</span> <span class="n">parent_1</span><span class="p">[</span><span class="n">np</span><span class="o">.</span><span class="n">sort</span><span class="p">(</span><span class="n">random_sequence</span><span class="p">)]</span>
</code></pre></div></div>

<p>For the last crossover method, we’ll switch it up a little bit with the <strong>Arithmetic</strong> crossover. Like the name implies, rather than swapping genes to form a new chromosome, we will do some arithmetics to make a new chromosome. We will perform a simple weighted average on the chromosomes, where the weight is randomly generated.</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/arithmetic_crossover.png" alt="alt" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">random_weight</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">rand</span><span class="p">()</span>
  <span class="n">child</span> <span class="o">=</span> <span class="n">parent_1</span> <span class="o">*</span> <span class="n">random_weight</span> <span class="o">+</span> <span class="n">parent_2</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">random_weight</span><span class="p">)</span>
</code></pre></div></div>

<p>Personally, I like using all of the crossover methods, so each time my algorithm performs crossover I randomly select one of the above methods with equal probability.</p>

<p>When setting up a genetic algorithm we define a probability of performing crossover, <script type="math/tex">p_\text{cross}</script>. Thus, with <script type="math/tex">1 - p_\text{cross}</script> probability, we carry over the parent chromosomes to the next generation without crossover. Since we are going to use elitism in our algorithm, we will probably want to set <script type="math/tex">p_\text{cross}</script> to be close to 1 because otherwise there is a high probability that we will have duplicate chromosomes in the next generation.</p>

<h3 id="mutation">Mutation</h3>

<p>After reviewing some of the crossover methods, you might be thinking that we’re just combining genes together without changing their order (with the exception of the arithmetic operator). This means that our chromosomes will be bounded by the initialized values from the first generation, which limits how much our agents can evolve. To ensure this doesn’t happen, we need to maintain genetic diversity - we do this with the mutation operator.</p>

<p>Similar to crossover, there are multiple ways to perform mutation. For my implementation I randomly select a gene with <script type="math/tex">p_\text{mutate}</script> probability and add gaussian noise to it:</p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    <span class="n">noise</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">standard_normal</span><span class="p">()</span> <span class="o">*</span> <span class="n">noise_scale</span>
    <span class="n">mutation_position</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">population</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
    <span class="n">child</span><span class="p">[</span><span class="n">mutation_position</span><span class="p">]</span> <span class="o">=</span> <span class="n">child</span><span class="p">[</span><span class="n">mutation_position</span><span class="p">]</span> <span class="o">+</span> <span class="n">noise</span>
</code></pre></div></div>

<p>Even though I remained relatively simple with my implementation, you can get a bit fancier by implementing some of the mutation methods outlined below. The first is the <strong>Swap</strong> mutation, which selects two random positions in the chromosome and swaps their genes:</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/swap_mutation.png" alt="alt" style="max-width: 300px; height: auto;" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">random_positions</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">child</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]),</span> <span class="mi">2</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
  <span class="n">value_1</span><span class="p">,</span> <span class="n">value_2</span> <span class="o">=</span> <span class="n">child</span><span class="p">[</span><span class="n">random_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">child</span><span class="p">[</span><span class="n">random_positions</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span>
  <span class="n">child</span><span class="p">[</span><span class="n">random_positions</span><span class="p">[</span><span class="mi">0</span><span class="p">]],</span> <span class="n">child</span><span class="p">[</span><span class="n">random_positions</span><span class="p">[</span><span class="mi">1</span><span class="p">]]</span> <span class="o">=</span> <span class="n">value_2</span><span class="p">,</span> <span class="n">value_1</span>
</code></pre></div></div>

<p>Another method you can implement is the <strong>Inversion</strong> mutation, which selects two random positions and inverts/reverses the substring of genes between them:</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/inversion_mutation.png" alt="alt" style="max-width: 300px; height: auto;" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">lower_limit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">child</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
  <span class="n">upper_limit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">lower_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">child</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
  <span class="n">child</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">child</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">][::</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span>
</code></pre></div></div>

<p>Lastly, you can implement the <strong>Scramble</strong> mutation, which selects two random positions and scrambles the positions of the genes within them:</p>

<p style="text-align: center;"><img src="http://localhost:4000/images/scramble_mutation.png" alt="alt" style="max-width: 300px; height: auto;" /></p>

<div class="language-python highlighter-rouge"><div class="highlight"><pre class="highlight"><code>  <span class="n">lower_limit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="n">child</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">]</span><span class="o">-</span><span class="mi">1</span><span class="p">)</span>
  <span class="n">upper_limit</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">randint</span><span class="p">(</span><span class="n">lower_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">,</span> <span class="n">child</span><span class="o">.</span><span class="n">shape</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
  <span class="n">scrambled_order</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="n">lower_limit</span><span class="p">,</span> <span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">),</span> <span class="n">upper_limit</span> <span class="o">+</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">lower_limit</span><span class="p">,</span> <span class="n">replace</span> <span class="o">=</span> <span class="bp">False</span><span class="p">)</span>
  <span class="n">child</span><span class="p">[</span><span class="n">lower_limit</span><span class="p">:</span><span class="n">upper_limit</span><span class="o">+</span><span class="mi">1</span><span class="p">]</span> <span class="o">=</span> <span class="n">child</span><span class="p">[</span><span class="n">scrambled_order</span><span class="p">]</span>
</code></pre></div></div>

<h2 id="deepminds-control-suite">DeepMind’s Control Suite</h2>

<p>Great, now that we have all the pieces to make a genetic algorithm, let’s put them together to train the “cheetah” domain from DeepMind’s Control Suite. For those who are not familiar with the library, it is powered by the MuJoCo physics engine and provides you with an environment to train agents on a set of continuous control tasks. For our experiment we want the cheetah to learn how to run.</p>

<p>The thing that I really like about this library is that it has a standardized structure. For example, the library provides you with an observation of the environment and a reward for every action you take. The state observation for our domain task is a combination of the cheetah’s position and velocity. The reward, <script type="math/tex">r</script>, is a function of the forward velocity, <script type="math/tex">v</script>, up to a maximum of <script type="math/tex">10 m/s</script>:</p>

<script type="math/tex; mode=display">r(v) = max(0, min(v/10, 1))</script>

<p>We run each episode for 500 frames and calculate the fitness, <script type="math/tex">f</script>, as:</p>

<script type="math/tex; mode=display">f = \sum_{i=1}^{500}r_i</script>

<p>At each time step, our agent has to make 6 actions in parallel - the movement of each of its limbs. The action vector for our cheetah has the following property: <script type="math/tex">\boldsymbol{a} \in \mathcal{A} \equiv [-1,1]^{6}</script>. Thus, for our policy, we are going to use a neural network with a 6-dimensional <script type="math/tex">\tanh</script> output. We flatten all of the neural network weights to a one dimensional array in order to implement the crossover and mutation operators mentioned above. After the child chromosome is created, we reshape the weights to be used in a neural network for the next generation. Overall, we used 1000 generations with a population size of 40 to train our cheetah.</p>

<p>When the training process starts (Generation 1), we see that the cheetah doesn’t know how to move and end up falling backwards:</p>

<p style="text-align: center;"><img src="/images/cheetah_start.gif" alt="Alt Text" /></p>

<p>As training progresses (Generation 250), the cheetah learns how to run forward. However, we see that near the end of the episode it loses control of its stride and falls flat on its face:</p>

<p style="text-align: center;"><img src="/images/cheetah_mid.gif" alt="Alt Text" /></p>

<p>At the end of the training process (Generation 1000), we see that the cheetah learns how to run, while also maintaining its center of gravity during large strides:</p>

<p style="text-align: center;"><img src="/images/cheetah_end.gif" alt="Alt Text" /></p>

<p>Awesome, we did it!</p>

<h2 id="concluding-remarks">Concluding Remarks</h2>

<p>In this post we learned how genetic algorithms can be used to optimize parameters of a neural network for a continuous control task. In a future post we will explore an application where we mix genetic algorithms (derivative-free method) and policy gradients (derivative-based method) for better training.</p>

<div id="disqus_thread"></div>
<script>

/**
*  RECOMMENDED CONFIGURATION VARIABLES: EDIT AND UNCOMMENT THE SECTION BELOW TO INSERT DYNAMIC VALUES FROM YOUR PLATFORM OR CMS.
*  LEARN WHY DEFINING THESE VARIABLES IS IMPORTANT: https://disqus.com/admin/universalcode/#configuration-variables*/
/*
var disqus_config = function () {
this.page.url = http://localhost:4000;  // Replace PAGE_URL with your page's canonical URL variable
this.page.identifier = /genetic-algorithm; // Replace PAGE_IDENTIFIER with your page's unique identifier variable
};
*/
(function() { // DON'T EDIT BELOW THIS LINE
var d = document, s = d.createElement('script');
s.src = 'https://brandinho-github-io.disqus.com/embed.js';
s.setAttribute('data-timestamp', +new Date());
(d.head || d.body).appendChild(s);
})();
</script>

<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>


        
      </section>

      <footer class="page__meta">
        
        
  


  
  
  

  <p class="page__taxonomy">
    <strong><i class="fas fa-fw fa-tags" aria-hidden="true"></i> Tags: </strong>
    <span itemprop="keywords">
    
      
      
      <a href="/tags/#genetic-algorithm" class="page__taxonomy-item" rel="tag">Genetic Algorithm</a><span class="sep">, </span>
    
      
      
      <a href="/tags/#reinforcement-learning" class="page__taxonomy-item" rel="tag">Reinforcement Learning</a>
    
    </span>
  </p>




        
          <p class="page__date"><strong><i class="fas fa-fw fa-calendar-alt" aria-hidden="true"></i> Updated:</strong> <time datetime="2018-11-18T00:00:00-05:00">November 18, 2018</time></p>
        
      </footer>

      <section class="page__share">
  
    <h4 class="page__share-title">Share on</h4>
  

  <a href="https://twitter.com/intent/tweet?text=Learning+How+to+Run+with+Genetic+Algorithms%20http%3A%2F%2Flocalhost%3A4000%2Fgenetic-algorithm%2F" class="btn btn--twitter" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Twitter"><i class="fab fa-fw fa-twitter" aria-hidden="true"></i><span> Twitter</span></a>

  <a href="https://www.facebook.com/sharer/sharer.php?u=http%3A%2F%2Flocalhost%3A4000%2Fgenetic-algorithm%2F" class="btn btn--facebook" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Facebook"><i class="fab fa-fw fa-facebook" aria-hidden="true"></i><span> Facebook</span></a>

  <a href="https://plus.google.com/share?url=http%3A%2F%2Flocalhost%3A4000%2Fgenetic-algorithm%2F" class="btn btn--google-plus" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on Google Plus"><i class="fab fa-fw fa-google-plus" aria-hidden="true"></i><span> Google+</span></a>

  <a href="https://www.linkedin.com/shareArticle?mini=true&url=http%3A%2F%2Flocalhost%3A4000%2Fgenetic-algorithm%2F" class="btn btn--linkedin" onclick="window.open(this.href, 'window', 'left=20,top=20,width=500,height=500,toolbar=1,resizable=0'); return false;" title="Share on LinkedIn"><i class="fab fa-fw fa-linkedin" aria-hidden="true"></i><span> LinkedIn</span></a>
</section>


      
  <nav class="pagination">
    
      <a href="/bayesian-policy/" class="pagination--pager" title="Learning Probability Distributions in Bounded Action Spaces
">Previous</a>
    
    
      <a href="/mario-ppo/" class="pagination--pager" title="Playing Super Mario Bros with Proximal Policy Optimization
">Next</a>
    
  </nav>

    </div>

    
  </article>

  
  
    <div class="page__related">
      <h4 class="page__related-title">You may also enjoy</h4>
      <div class="grid__wrapper">
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/mario-ppo/" rel="permalink">Playing Super Mario Bros with Proximal Policy Optimization
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  16 minute read
</p>
    
    <p class="archive__item-excerpt" itemprop="description">Reinforcement Learning, Neural Networks, Policy Gradient
</p>
  </article>
</div>
        
          



<div class="grid__item">
  <article class="archive__item" itemscope itemtype="http://schema.org/CreativeWork">
    
    <h2 class="archive__item-title" itemprop="headline">
      
        <a href="/bayesian-policy/" rel="permalink">Learning Probability Distributions in Bounded Action Spaces
</a>
      
    </h2>
    
      <p class="page__meta"><i class="far fa-clock" aria-hidden="true"></i> 




  11 minute read
</p>
    
    <p class="archive__item-excerpt" itemprop="description">Reinforcement Learning, Neural Networks, Bayesian
</p>
  </article>
</div>
        
      </div>
    </div>
  
  
</div>
    </div>

    

    <div class="page__footer">
      <footer>
        <!-- start custom footer snippets -->

<!-- end custom footer snippets -->
        <div class="page__footer-follow">
  <ul class="social-icons">
    
      <li><strong>Follow:</strong></li>
    

    
      
        
      
        
      
        
      
        
      
        
      
        
      
    

    <li><a href="/feed.xml"><i class="fas fa-fw fa-rss-square" aria-hidden="true"></i> Feed</a></li>
  </ul>
</div>

<div class="page__footer-copyright">&copy; 2018 Brandon Da Silva. Powered by <a href="https://jekyllrb.com" rel="nofollow">Jekyll</a> &amp; <a href="https://mademistakes.com/work/minimal-mistakes-jekyll-theme/" rel="nofollow">Minimal Mistakes</a>.</div>

      </footer>
    </div>

    
  <script src="/assets/js/main.min.js"></script>
  <script src="https://use.fontawesome.com/releases/v5.3.1/js/all.js"></script>








<script type="text/javascript" async
  src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-MML-AM_CHTML">
</script>



  </body>
</html>
